### PR TITLE
PLANNER-2470: Add Benchmarker Metrics to Solver via Micrometer

### PR DESCRIPTION
**Pull request by Christopher-Chianelli**
Note: new implementation in benchmark is poll (i.e. measure every 5 seconds) based instead of step (i.e. measure at end of every step) based. Still need to add tests for new metrics.

### JIRA

https://issues.redhat.com/browse/PLANNER-2470

### Checklist
- [ ] Documentation updated if applicable.
- [ ] Upgrade recipe provided if applicable.

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* for a <b>pull request build</b> please add comment: <b>Jenkins retest this</b>
* for a <b>specific pull request build</b> please add comment: <b>Jenkins (re)run [optaplanner|apps|examples] tests</b>
* for a <b>full downstream build</b> please add comment: <b>Jenkins run fdb</b>
* for a <b>compile downstream build</b> please add comment: <b>Jenkins run cdb</b>
* for a <b>full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>
* for an <b>upstream build</b> please add comment: <b>Jenkins run upstream</b>
* for a <b>Quarkus LTS check</b> please add comment: <b>Jenkins run LTS</b>
* for a <b>specific Quarkus LTS check</b> please add comment: <b>Jenkins (re)run [optaplanner|apps|examples] LTS</b>
* for a <b>Native check</b> please add comment: <b>Jenkins run native</b>
* for a <b>specific Native LTS check</b> please add comment: <b>Jenkins (re)run [optaplanner|apps|examples] native</b>
</details>

### CODE REVIEW

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java
Comment (triceo) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java:43]: Let's make this generic in `Solution_` and get rid of all the raw type warnings in this class.

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java
Comment (triceo) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java:48]: I often point this out in your PRs. Please make sure fields are **always `private` and always `final`**, unless there is a reason for them not to be.

This should apply not only to this PR, but to all future PRs, please.

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java
Comment (triceo) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java:62]: This works better as a property of the `SolverMetric` directly.

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java
Comment (triceo) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java:86]: In this fluent chain, I find it hard to see where micrometer ends and streams start.
May be worth splitting former into its own variable.
Also, `Meter::getId`.

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java
Comment (triceo) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java:100]: This is potentially a very expensive operation, and it is being used often.
I'd consider a form of cache, at the very least for the `find()` result.

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java
Comment (triceo) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java:108]: Dtto. We do not need to look them up all the time, we can gradually build a cache.
(I'm of course assuming that the references do not change.)

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/bestsolutionmutation/BestSolutionMutationSubSingleStatistic.java
Comment (triceo) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/bestsolutionmutation/BestSolutionMutationSubSingleStatistic.java:52]: Shouldn't we unregister the listeners here? If so, this needs to happen in all the statistics.
Or is that something that is not supported by micrometer?

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/movecountperstep/MoveCountPerStepSubSingleStatistic.java
Comment (triceo) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/movecountperstep/MoveCountPerStepSubSingleStatistic.java:46]: I don't understand. Why the nesting? The gauge is the same, shouldn't therefore `accepted` and `selected` do the same thing?

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/scorecalculationspeed/ScoreCalculationSpeedSubSingleStatistic.java
Comment (triceo) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/scorecalculationspeed/ScoreCalculationSpeedSubSingleStatistic.java:34]: This appears unused. Maybe some code is missing?

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/subsingle/constraintmatchtotalbestscore/ConstraintMatchTotalBestScoreSubSingleStatistic.java
Comment (triceo) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/subsingle/constraintmatchtotalbestscore/ConstraintMatchTotalBestScoreSubSingleStatistic.java:107]: This is complex. It deserves some comments on what does it do and why.

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/subsingle/constraintmatchtotalstepscore/ConstraintMatchTotalStepScoreSubSingleStatistic.java
Comment (triceo) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/subsingle/constraintmatchtotalstepscore/ConstraintMatchTotalStepScoreSubSingleStatistic.java:106]: Dtto.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/DefaultLocalSearchPhase.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/DefaultLocalSearchPhase.java:154]: Nothing for construction heuristics?
That may be why I was seeing the empty graphs.

We need to figure out how to make this happen for every phase.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/statistic/BestScoreStatistic.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/statistic/BestScoreStatistic.java:26]: Instead of implementing `Consumer<Solver>`, which is a generic interface, I'd create our own interface.
This is a very specific functionality and we want to see where it is used in the code. `Consumer<Solver>` doesn't let us do this, but `SolverStatistic<Solver>` would.

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/memoryuse/MemoryUseSubSingleStatistic.java
Comment (triceo) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/memoryuse/MemoryUseSubSingleStatistic.java:79]: Is `max` what we want? I'd expect we want the current usage, not the most it's even been.

**Review by triceo — CHANGES_REQUESTED** [review]
Good first pass. See DM on Zulip for a serious issue.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java:88]: Let's not expose it in the solver config just yet. Let's give it a while to settle down.

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/bestsolutionmutation/BestSolutionMutationSubSingleStatistic.java
Comment (Christopher-Chianelli) [reply optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/bestsolutionmutation/BestSolutionMutationSubSingleStatistic.java:52]: The `StatisticRegistry` itself is unregistered/removed from `Metrics`. See `SubSingleBenchmarkRunner`:

```
        StatisticRegistry statisticRegistry = new StatisticRegistry((DefaultSolver) solver);
        Metrics.addRegistry(statisticRegistry);
        solver.addEventListener(statisticRegistry);
        solver.addPhaseLifecycleListener(statisticRegistry);

        Tags runTag = Tags.of("optaplanner.benchmark.run", runId);
        for (SubSingleStatistic<Solution_, ?> subSingleStatistic : subSingleBenchmarkResult.getEffectiveSubSingleStatisticMap()
                .values()) {
            subSingleStatistic.open(statisticRegistry, runTag, solver);
            subSingleStatistic.initPointList();
        }
        Solution_ solution = solver.solve(problem);

        solver.removePhaseLifecycleListener(statisticRegistry);
        solver.removeEventListener(statisticRegistry);
        Metrics.removeRegistry(statisticRegistry);
```

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/movecountperstep/MoveCountPerStepSubSingleStatistic.java
Comment (Christopher-Chianelli) [reply optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/movecountperstep/MoveCountPerStepSubSingleStatistic.java:46]: The guage should be different (it should be the selected gauge); good catch

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/bestsolutionmutation/BestSolutionMutationSubSingleStatistic.java
Comment (triceo) [reply optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/bestsolutionmutation/BestSolutionMutationSubSingleStatistic.java:52]: In that case, I wouldn't override these methods. Make it default, add a comment in there.

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/memoryuse/MemoryUseSubSingleStatistic.java
Comment (Christopher-Chianelli) [reply optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/memoryuse/MemoryUseSubSingleStatistic.java:79]: MemoryUseStatisticPoint use both current and max

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/DefaultLocalSearchPhase.java
Comment (Christopher-Chianelli) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/DefaultLocalSearchPhase.java:154]: A bunch of metrics in the original benchmarker code is set up for local search only (but data collection is unrelated to it; should be able to make it every step)

Comment (Christopher-Chianelli) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/DefaultLocalSearchPhase.java:154]: Looking at the class it on, it for the statistics that are local search specific

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java
Comment (triceo) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java:55]: Aren't you missing BigDecimal here?

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java
Comment (triceo) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java:114]: I'm thinking... shouldn't you be able to just build the score from all its levels, and then retrieve the level numbers from it?

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/movecountperstep/MoveCountPerStepSubSingleStatistic.java
Comment (triceo) [inline optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/movecountperstep/MoveCountPerStepSubSingleStatistic.java:46]: Since you control the registry API, I suggest you continue accepting the enum and do the `getMeterId()` call internally. Less repetition.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java:651]: How do we make sure that this is not exposed to the users?
(Geoffrey wants to keep in control of the solver config. But he will be gone for 3 weeks. I'd like to merge this earlier, but I won't be able to, if it impacts public API.)

File: optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java
Comment (Christopher-Chianelli) [reply optaplanner-benchmark/src/main/java/org/optaplanner/benchmark/impl/statistic/StatisticRegistry.java:114]: Score definitions expect the score levels to be a particular type (Exactly integer, exactly long, etc.). Micrometer gauges can only return doubles. So in order to not get a typecast exception in the score definition, I need to convert it to the "expected" type. (There are micrometer gauges for each level)

File: optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java
Comment (Christopher-Chianelli) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java:651]: Currently, it XML transitive (i.e. can be accessed via Java API, but cannot via XML API. It need to be in the config so it can be read by the SolverFactory (which uses it to set up metrics)

Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java:651]: The problem with it being in the Java API is that it is a public API that needs to be kept stable.
Therefore I can not merge this PR until discussed with and approved by Geoffrey.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/statistic/SolverStatistic.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/statistic/SolverStatistic.java:5]: This could be generic in `Solution_`, therefore eliminating some of the unchecked warnings you are suppressing.

**Comment by triceo** [comment]
Since we'll be doing the API review as part of this PR, please reintroduce the solver config changes to affect the XML, @Christopher-Chianelli .

**Comment by Christopher-Chianelli** [comment]
`MetricConfig -> MonitoringConfig`, tag map now part of internal DefaultSolver API with default tags set in SolverManager and Benchmarker (no default id tag set in default solver factory)

**Comment by triceo** [comment]
What's missing to mark this PR as ready for review, @Christopher-Chianelli ?

**Comment by Christopher-Chianelli** [comment]
@triceo strictly speaking, no new test coverage, although how to test metrics beside "they have a value" is beyond me. However, it does have about the same test coverage for metrics as the benchmarker (no tests were removed/modified).

**Comment by triceo** [comment]
Well, if we spend some extra time, there are things we could test that are useful. If the scenarios are defined well enough, with a step count termination, you should be able to know the expected values for some of these metrics, such as score calc count. You can test solve lengths and solver counts.

I think it would be worth to invest that time. We definitely don't need to test all the metrics; for example, mutations per step seems excessive. But I worry that if we say we do not test metrics in principle, that's just going to bring technical debt straight into core. 

The fact that benchmarker is largely untested is a problem that we've been lamenting for years, and it makes the code hard to touch. And that makes people generally avoid it.

**Comment by Christopher-Chianelli** [comment]
Solver count and error count is already tested (they were added w/ tests in a previous PR). If you can give me configs with expected metric values, I can write tests for them (although I can only really test "final" values for some metrics since some metrics use EventListeners, and I cannot guarantee the order of event listeners.

**Comment by triceo** [comment]
Case in point: best score metrics. You can create a solver with a single entity, single variable and only a handful of values. You can set the solver env mode to reproducible, therefore you will know what's the expected best score at each step. And you can write a test to check that this is reflected in the metrics.

This approach is not feasible for some of the other metrics, but for best score, it should be relatively simple, I think.

Also, there appear to be conflicts now. I do not want to hold up merging the PR anymore; if you don't see how you can do a test like I describe, we can discuss later.

**Comment by Christopher-Chianelli** [comment]
@triceo added test for best score metric

**Review by triceo — APPROVED** [review]
Personally, I think that this step-based checking makes the test relatively unclear, without bringing too much benefit. How about we only check the metric at the end of solving?

After that, I consider this PR good to merge.

**Comment by Christopher-Chianelli** [comment]
@triceo I disagree; checking only at the end of solving won't catch bugs relating to only updating once / only updating at end, which would make the benchmark best score graph a lot less valuable.

**Comment by triceo** [comment]
@Christopher-Chianelli There appears to be a Windows-specific issue in the test.
I see your point of needing to check after every step.

**Comment by Christopher-Chianelli** [comment]
@triceo fixed the Windows issue; it seems a gauge existed but returned NaN, so I added `Double.isFinite()` checks to the registries when the values are recorded

**Pull request merged**
